<a href="https://colab.research.google.com/github/ges45-learn/Module5/blob/main/SCS_Colab_Notebook_W5_v0.2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Course notebook notes

Throughout this notebook you will be taught how to work with Colab and OpenLane. More specifically, you will use this Notebook to:

* compare IC designs from 1970s and 2020s  
* explore a basic semiconductor layout, identifying the materials that are included in it
* study and characterise a given standard library cell.
* look at a design tool, explore it and consider what it is doing
* look at a design tool, inserting input/output (I/O) pads for wiring bonding and experience verification

**NOTE 1:** In module 1, we asked you to create a Google and Colab account. For optimal experience, you must be logged into the Google account that you created every time you work with this file.

**NOTE 2:** If you have not done so already, be sure to make a copy of this notebook in your drive. This will help you retain any changes you make and ensure a smooth journey through the following sections.

---

# Imports, authentication and preparation
Running the following cells will import all the data we need for this notebook to work. Simply select the code cell below and it will import all of these important packages.

Note that you will need to run these every time you open this notebook. Be sure to complete this step before attempting any of the tasks in this notebook.

Setting up the OpenLane runtime environment

For this to work, we want to:
* install OpenLane (v2) and its dependencies
* download and set up [open source sky130 PDK](https://github.com/google/skywater-pdk/) by Google and Skywater.

Created by Efabless, *edited by Dr Matthew Tang*.


In [1]:
# @title Setup Nix {display-mode: "form"}
# @markdown Nix is a package manager with an emphasis on reproducible builds,
# @markdown     and it is the primary method for installing OpenLane2.
# @markdown
# @markdown     This step installs the Nix package manager and enables the
# @markdown     experimental 'flakes' feature.
# @markdown
# If you're not in a Colab, this just sets the environment variables.
# You will need to install Nix and enable flakes on your own following
# [this guide](https://openlane2.readthedocs.io/en/stable/getting_started/common/nix_installation/index.html).
import os
import sys
import shutil

os.environ["LOCALE_ARCHIVE"] = "/usr/lib/locale/locale-archive"

if "google.colab" in sys.modules:
    if shutil.which("nix-env") is None:
        !curl -L https://nixos.org/nix/install | bash -s -- --daemon --yes
        !echo "extra-experimental-features = nix-command flakes" >> /etc/nix/nix.conf
        !killall nix-daemon
else:
    if shutil.which("nix-env") is None:
        raise RuntimeError("Nix is not installed!")

os.environ["PATH"] = f"/nix/var/nix/profiles/default/bin/:{os.getenv('PATH')}"

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  4499  100  4499    0     0  13972      0 --:--:-- --:--:-- --:--:-- 13972
downloading Nix 2.35.2 binary tarball for x86_64-linux from 'https://releases.nixos.org/nix/nix-2.35.2/nix-2.35.2-x86_64-linux.tar.xz' to '/tmp/nix-binary-tarball-unpack.0LY7WTxYGJ'...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 25.8M  100 25.8M    0     0   132M      0 --:--:-- --:--:-- --:--:--  132M
Note: a multi-user installation is possible. See https://nix.dev/manual/nix/stable/installation/installing-binary.html#multi-user-installation
Switching to the Multi-user Installer
Welcome to the Multi-User Nix Installation

This installation tool will set up yo

In [2]:
# @title Get OpenLane2 and SKY130 Open PDK {"display-mode":"form"}
# @markdown Click the ▷ button to download and install OpenLane.
# @markdown
# @markdown     This will install OpenLane's tool dependencies using Nix,
# @markdown     and OpenLane itself using PIP.
# @markdown
# @markdown     Note that `python3-tk` may need to be installed using your OS's
# @markdown     package manager.
import os
import subprocess
import IPython

openlane_version = "main"  # @param {key:"OpenLane Version", type:"string"}

if openlane_version == "latest":
    openlane_version = "main"

pdk_root = "~/.volare"  # @param {key:"PDK Root", type:"string"}

pdk_root = os.path.expanduser(pdk_root)

pdk = "sky130"  # @param {key:"PDK (without the variant)", type:"string"}

openlane_ipynb_path = os.path.join(os.getcwd(), "openlane_ipynb")

display(IPython.display.HTML("<h3>Downloading OpenLane…</a>"))

TESTING_LOCALLY = False
!rm -rf {openlane_ipynb_path}
!mkdir -p {openlane_ipynb_path}
if TESTING_LOCALLY:
    !ln -s {os.getcwd()} {openlane_ipynb_path}
else:
    !curl -L "https://github.com/efabless/openlane2/tarball/{openlane_version}" | tar -xzC {openlane_ipynb_path} --strip-components 1

try:
    import tkinter
except ImportError:
    if "google.colab" in sys.modules:
        !sudo apt-get install python-tk

try:
    import tkinter
except ImportError as e:
    display(
        IPython.display.HTML(
            '<h3 style="color: #800020";>❌ Failed to import the <code>tkinter</code> library for Python, which is required to load PDK configuration values. Make sure <code>python3-tk</code> or equivalent is installed on your system.</a>'
        )
    )
    raise e from None


display(IPython.display.HTML("<h3>Downloading OpenLane's dependencies…</a>"))
try:
    subprocess.check_call(
        ["nix", "profile", "install", ".#colab-env", "--accept-flake-config"],
        cwd=openlane_ipynb_path,
    )
except subprocess.CalledProcessError as e:
    display(
        IPython.display.HTML(
            '<h3 style="color: #800020";>❌ Failed to install binary dependencies using Nix…</h3>'
        )
    )

display(IPython.display.HTML("<h3>Downloading Python dependencies using PIP…</a>"))
try:
    subprocess.check_call(
        ["pip3", "install", "."],
        cwd=openlane_ipynb_path,
    )
except subprocess.CalledProcessError as e:
    display(
        IPython.display.HTML(
            '<h3 style="color: #800020";>❌ Failed to install Python dependencies using PIP…</h3>'
        )
    )
    raise e from None

display(IPython.display.HTML("<h3>Downloading PDK…</a>"))

#pdk_tmp = user_workspace_path + '/.volare'
#if os.path.exists(pdk_tmp):
#  IPython.display.HTML('<p>PDK found in user workspace. Linking ...</p>')


import volare

volare.enable(
    volare.get_volare_home(pdk_root),
    pdk,
    open(
        os.path.join(openlane_ipynb_path, "openlane", "open_pdks_rev"),
        encoding="utf8",
    )
    .read()
    .strip(),
)
# create symlink for PDK
#!ln -s {pdk_root} ~/.volare
sys.path.insert(0, openlane_ipynb_path)

display(IPython.display.HTML("<h3>⭕️ Done.</a>"))

import logging

# Remove the stupid default colab logging handler
logging.getLogger().handlers.clear()

# link the latest versions with the assumed directory name
!cd /root/.volare/volare/sky130/versions && ln -s `cat ../current` bdc9412b3e468c102d01b7cf6337be06ec6e9c9a

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 9527k    0 9527k    0     0  11.4M      0 --:--:-- --:--:-- --:--:-- 11.4M


Version 0fe599b2afb6708d281543108caf8310912f54af not found locally, attempting to download…

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Version 0fe599b2afb6708d281543108caf8310912f54af enabled for the sky130 PDK.

## Quick check

After running the above cells, you should see the version number of OpenLane. If you see that, you are ready to run the tasks below.

In [3]:
import openlane
print(openlane.__version__)

2.3.10


## Utility functions

These functions are required to generate high-quality images from the OpenLane's outputs (GDSII layout). Execute the code box once before beginning any tasks in the course.

[1] Create images from GDSII layout (`gds_render.py`).

This short Python program generates a high-resolution image from the GDSII physical layout provided, then inspect the images closely in your browser. (You may download the image too).

Recommended resolution:

Module 1: 2,000 pixels

Other modules: 1,000 pixels

In [4]:
%%writefile /content/gds_render.py
# @markdown You may adjust the resolution of the image by modifying the value below. The default is 1,000 pixels.
#!/usr/bin/env python3
# Copyright (c) 2021-2022 Efabless Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Original Copyright Follows
#
# BSD 3-Clause License
#
# Copyright (c) 2018, The Regents of the University of California
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice, this
#   list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
#
# * Neither the name of the copyright holder nor the names of its
#   contributors may be used to endorse or promote products derived from
#   this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import pya
import sys

def render(
    input: str,
    output: str,
):
    PDK = '~/.volare/volare/sky130/versions/bdc9412b3e468c102d01b7cf6337be06ec6e9c9a'
    lyt = f'{PDK}/sky130A/libs.tech/klayout/tech/sky130A.lyt'
    lyp = f'{PDK}/sky130A/libs.tech/klayout/tech/sky130A.lyp'
    lym = f'{PDK}/sky130A/libs.tech/klayout/tech/sky130A.lyt'

    img_res = 1000 # @param {key:"image resolution", type:"number"}
    print(f"image resolution: {img_res}")
    # display ONLY the layers in the following list
    selected_layers = [
        "diff.",
        "dnwell.",
        "li1.",
        "licon1.",
        "mcon.",
        "met1.",
        "met2.",
        "met3.",
        "met4.",
        "met5.",
        #"vnpc.",
        #"vnsdm.",
        #"vnwell.",
        "poly.",
        "psdm.",
        "pwell.",
        "tap.",
        "via.",
        "via2.",
        "via3.",
        "via4."
    ]

    try:
        gds = input.endswith(".gds")

        # Load technology file
        tech = pya.Technology()
        tech.load(lyt)

        layout_options = None

        view = pya.LayoutView()
        view.load_layer_props(lyp)

        if gds:
            view.load_layout(input)
        else:
            view.load_layout(input, layout_options, lyt)

        view.max_hier()

        # control the visibility of the layers
        for layer in view.each_layer():
            layer.visible = False
            if '122/16@1' in layer.source: # grid and ruler
                layer.visible = True
            for t in selected_layers:
                if layer.name.startswith(t):
                    layer.visible = True

        pixels = view.get_pixels_with_options(img_res, img_res)

        with open(output, "wb") as f:
            f.write(pixels.to_png_data())

    except Exception as e:
        print(e)
        exit(1)


if __name__ == "__main__":
    gds_file = ''
    if len(sys.argv) >= 2:
        gds_file = sys.argv[1]
    else:
        quit('Please provide a GDS file.')

    out_file = 'output.png'
    if len(sys.argv) >= 3:
        out_file = sys.argv[2]
    print(f"Input GDS file: {gds_file}")
    print(f"Output PNG file: {out_file}")

    render(gds_file, out_file)
    print("done.")

Writing /content/gds_render.py


[2] Install a customised render function.

In [5]:
%%writefile /content/openlane_ipynb/openlane/scripts/klayout/render.py
#!/usr/bin/env python3
# @markdown This makes ``display()`` to generate an additional png image (``/content/openlane_run/output.png``) of a higher resolution (default: 5000 pixels).
# More information about layers can be found at this
# [link](https://skywater-pdk.readthedocs.io/en/main/rules/layers.html). To hide the layer in the output image, comment off from the list.
#
# Copyright (c) 2021-2022 Efabless Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Tuple
import pya
import click
@click.command()
@click.option("-o", "--output", required=True)
@click.option("-l", "--input-lef", "input_lefs", multiple=True,)
@click.option( "-T", "--lyt", required=True, help="KLayout .lyt file",)
@click.option( "-P", "--lyp", required=True, help="KLayout .lyp file",)
@click.option( "-M", "--lym", required=True, help="KLayout .map (LEF/DEF layer map) file",)
@click.argument("input")
def render(input_lefs: Tuple[str, ...], output: str, lyt: str, lyp: str, lym: str, input: str,):
    image_res = 5000 # output image resolution = 5000 pixels
    # display ONLY the layers in the following list
    selected_layers = [
        "diff.",
        "dnwell.",
        "li1.",
        "licon1.",
        "mcon.",
        "met1.",
        "met2.",
        "met3.",
        "met4.",
        "met5.",
        #"vnpc.",
        #"vnsdm.",
        #"vnwell.",
        "poly.",
        "psdm.",
        "pwell.",
        "tap.",
        "via.",
        "via2.",
        "via3.",
        "via4."
    ]

    try:
        gds = input.endswith(".gds")

        # Load technology file
        tech = pya.Technology()
        tech.load(lyt)

        layout_options = None
        if not gds:
            layout_options = tech.load_layout_options
            layout_options.lefdef_config.map_file = lym
            layout_options.lefdef_config.macro_resolution_mode = 1
            layout_options.lefdef_config.read_lef_with_def = False
            layout_options.lefdef_config.lef_files = list(input_lefs)

        view = pya.LayoutView()
        view.load_layer_props(lyp)

        if gds:
            view.load_layout(input)
        else:
            view.load_layout(input, layout_options, lyt)

        view.max_hier()

        # control the visibility of the layers
        for layer in view.each_layer():
            layer.visible = False
            if '122/16@1' in layer.source: # grid and ruler
                layer.visible = True
            for t in selected_layers:
                if t in layer.name:
                    layer.visible = True

        pixels = view.get_pixels_with_options(1000, 1000)

        with open(output, "wb") as f:
            f.write(pixels.to_png_data())

        # save an image of larger resolution
        pixels = view.get_pixels_with_options(image_res, image_res)
        with open("/content/openlane_run/output.png", "wb") as f:
            f.write(pixels.to_png_data())

    except Exception as e:
        print(e)
        exit(1)

if __name__ == "__main__":
    render()

Overwriting /content/openlane_ipynb/openlane/scripts/klayout/render.py


Now you may proceed with the actual task for the module.


---



# Module 5: Assembly, testing and packaging
In this section you will work on your fifth module task. Before attempting it, please ensure that you are working in your own copy of the notebook and you have successfully run all the cells above.

In this task, we are going to prepare an inverter design for assembly and packaging using the popular wire bonding technology.

---

**What you need to do**

Firstly, set up Nix, OpenLane2 and SKY130 Open PDK in your Colab runtime environment, and then action the following steps:
1. Write the design file and install the utility function.
1. Run automatic floorplanning.
1. Run manual floorplanning with fixed die size.
1. Work out the expected die size with I/O pads inserted.

The following sections contain the code and packages you need to complete these steps.

## Step 1: Write the design file

Run the following code to create a verilog file for a 32-bit inverter.

In [6]:
%%writefile /content/inverter32.v
module inverter32 (
    input  wire [31:0] a_in,
    output wire [31:0] y_out
);

    assign y_out = ~a_in;
endmodule

Writing /content/inverter32.v


Run the next box to install a utility function to visualise the layout.

In [7]:
%%writefile /content/openlane_ipynb/openlane/scripts/klayout/render.py
#!/usr/bin/env python3
# @markdown This makes ``display()`` to generate image of a higher resolution. Only the metal layers are shown (others are hidden in this task).
# More information about layers can be found at this
# [link](https://skywater-pdk.readthedocs.io/en/main/rules/layers.html). To hide the layer in the output image, comment off from the list.
#
# Copyright (c) 2021-2022 Efabless Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Tuple
import pya
import click
@click.command()
@click.option("-o", "--output", required=True)
@click.option("-l", "--input-lef", "input_lefs", multiple=True,)
@click.option( "-T", "--lyt", required=True, help="KLayout .lyt file",)
@click.option( "-P", "--lyp", required=True, help="KLayout .lyp file",)
@click.option( "-M", "--lym", required=True, help="KLayout .map (LEF/DEF layer map) file",)
@click.argument("input")
def render(input_lefs: Tuple[str, ...], output: str, lyt: str, lyp: str, lym: str, input: str,):
    image_res = 1000 # @param {key:"image resolution", type:"number"}
    # display ONLY the layers in the following list
    selected_layers = [
        #"diff.",
        #"dnwell.",
        #"li1.",
        #"licon1.",
        #"mcon.",
        "met1.",
        "met2.",
        "met3.",
        "met4.",
        "met5.",
        #"vnpc.",
        #"vnsdm.",
        #"vnwell.",
        #"poly.",
        #"psdm.",
        #"pwell.",
        #"tap.",
        "via.",
        "via2.",
        "via3.",
        "via4."
    ]

    try:
        gds = input.endswith(".gds")

        # Load technology file
        tech = pya.Technology()
        tech.load(lyt)

        layout_options = None
        if not gds:
            layout_options = tech.load_layout_options
            layout_options.lefdef_config.map_file = lym
            layout_options.lefdef_config.macro_resolution_mode = 1
            layout_options.lefdef_config.read_lef_with_def = False
            layout_options.lefdef_config.lef_files = list(input_lefs)

        view = pya.LayoutView()
        view.load_layer_props(lyp)

        if gds:
            view.load_layout(input)
        else:
            view.load_layout(input, layout_options, lyt)

        view.max_hier()

        # control the visibility of the layers
        for layer in view.each_layer():
            layer.visible = False
            if '122/16@1' in layer.source: # grid and ruler
                layer.visible = True
            for t in selected_layers:
                if t in layer.name:
                    layer.visible = True

        pixels = view.get_pixels_with_options(image_res, image_res)

        with open(output, "wb") as f:
            f.write(pixels.to_png_data())

        # save an image of larger resolution
        pixels = view.get_pixels_with_options(image_res, image_res)
        with open("/content/openlane_run/output.png", "wb") as f:
            f.write(pixels.to_png_data())

    except Exception as e:
        print(e)
        exit(1)

if __name__ == "__main__":
    render()

Overwriting /content/openlane_ipynb/openlane/scripts/klayout/render.py


## Step 2: First trial – automatic floorplan

We will set up OpenLane as usual, taking all default settings.

In [8]:
from openlane.steps import Step
from openlane.state import State
from openlane.config import Config

Config.interactive(
    "inverter32",
    PDK="sky130A",
    PRIMARY_GDSII_STREAMOUT_TOOL="klayout",
)


### Interactive Configuration
#### Initial Values

<br />

```yaml
CELL_BB_VERILOG_MODELS:
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/verilog/sky130_fd_sc_hd__blackbox.v
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/verilog/sky130_fd_sc_hd__blackbox_pp.v
CELL_GDS:
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/gds/sky130_fd_sc_hd.gds
CELL_LEFS:
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_fd_sc_hd.lef
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_ef_sc_hd.lef
CELL_PAD_EXCLUDE:
- sky130_fd_sc_hd__tap*
- sky130_fd_sc_hd__decap*
- sky130_ef_sc_hd__decap*
- sky130_fd_sc_hd__fill*
CELL_SPICE_MODELS:
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/spice/sky130_ef_sc_hd__decap_12.spice
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/spice/sky130_ef_sc_hd__fill_12.spice
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/spice/sky130_ef_sc_hd__fill_4.spice
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/spice/sky130_ef_sc_hd__fill_8.spice
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/spice/sky130_fd_sc_hd.spice
CELL_VERILOG_MODELS:
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/verilog/primitives.v
- /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/verilog/sky130_fd_sc_hd.v
CLOCK_NET: null
CLOCK_PERIOD: 10
CLOCK_PORT: null
CLOCK_TRANSITION_CONSTRAINT: 0.15
CLOCK_UNCERTAINTY_CONSTRAINT: 0.25
CLOCK_WIRE_RC_LAYERS: null
DECAP_CELL:
- sky130_ef_sc_hd__decap_12
- sky130_fd_sc_hd__decap_8
- sky130_fd_sc_hd__decap_6
- sky130_fd_sc_hd__decap_4
- sky130_fd_sc_hd__decap_3
DEFAULT_CORNER: nom_tt_025C_1v80
DEFAULT_MAX_TRAN: null
DESIGN_DIR: .
DESIGN_NAME: inverter32
DIE_AREA: null
DIODE_CELL: sky130_fd_sc_hd__diode_2/DIODE
ENDCAP_CELL: sky130_fd_sc_hd__decap_3
EXTRA_EXCLUDED_CELLS: null
EXTRA_GDS_FILES: null
EXTRA_LEFS: null
EXTRA_LIBS: null
EXTRA_SPICE_MODELS: null
EXTRA_VERILOG_MODELS: null
FALLBACK_SDC_FILE: /content/openlane_ipynb/openlane/scripts/base.sdc
FILL_CELL:
- sky130_fd_sc_hd__fill*
FP_IO_HLAYER: met3
FP_IO_VLAYER: met2
FP_TAPCELL_DIST: 13
FP_TRACKS_INFO: /root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/tracks.info
GND_NETS: null
GND_PIN: VGND
GPIO_PADS_LEF:
- /root/.volare/sky130A/libs.ref/sky130_fd_io/lef/sky130_fd_io.lef
- /root/.volare/sky130A/libs.ref/sky130_fd_io/lef/sky130_ef_io.lef
GPIO_PADS_LEF_CORE_SIDE:
- /root/.volare/sky130A/libs.tech/openlane/custom_cells/lef/sky130_fd_io_core.lef
- /root/.volare/sky130A/libs.tech/openlane/custom_cells/lef/sky130_ef_io_core.lef
GPIO_PADS_VERILOG:
- /root/.volare/sky130A/libs.ref/sky130_fd_io/verilog/sky130_ef_io.v
GPIO_PAD_CELLS:
- sky130_fd_io*
- sky130_ef_io*
IO_DELAY_CONSTRAINT: 20
LIB:
  '*_ff_n40C_1v95':
  - /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lib/sky130_fd_sc_hd__ff_n40C_1v95.lib
  '*_ss_100C_1v60':
  - /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lib/sky130_fd_sc_hd__ss_100C_1v60.lib
  '*_tt_025C_1v80':
  - /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lib/sky130_fd_sc_hd__tt_025C_1v80.lib
MACROS: null
MAX_CAPACITANCE_CONSTRAINT: 0.2
MAX_FANOUT_CONSTRAINT: 10
MAX_TRANSITION_CONSTRAINT: 0.75
OUTPUT_CAP_LOAD: 33.442
PDK: sky130A
PDK_ROOT: /root/.volare
PLACE_SITE: unithd
PNR_EXCLUDED_CELL_FILE: /root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/drc_exclude.cells
PRIMARY_GDSII_STREAMOUT_TOOL: klayout
RT_MAX_LAYER: met5
RT_MIN_LAYER: met1
SCL_GROUND_PINS:
- VGND
- VNB
SCL_POWER_PINS:
- VPWR
- VPB
SIGNAL_WIRE_RC_LAYERS: null
STA_CORNERS:
- nom_tt_025C_1v80
- nom_ss_100C_1v60
- nom_ff_n40C_1v95
- min_tt_025C_1v80
- min_ss_100C_1v60
- min_ff_n40C_1v95
- max_tt_025C_1v80
- max_ss_100C_1v60
- max_ff_n40C_1v95
STD_CELL_LIBRARY: sky130_fd_sc_hd
SYNTH_BUFFER_CELL: sky130_fd_sc_hd__buf_2/A/X
SYNTH_CLK_DRIVING_CELL: null
SYNTH_DRIVING_CELL: sky130_fd_sc_hd__inv_2/Y
SYNTH_EXCLUDED_CELL_FILE: /root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/no_synth.cells
SYNTH_TIEHI_CELL: sky130_fd_sc_hd__conb_1/HI
SYNTH_TIELO_CELL: sky130_fd_sc_hd__conb_1/LO
TECH_LEFS:
  max_*: /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/techlef/sky130_fd_sc_hd__max.tlef
  min_*: /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/techlef/sky130_fd_sc_hd__min.tlef
  nom_*: /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/techlef/sky130_fd_sc_hd__nom.tlef
TIME_DERATING_CONSTRAINT: 5
TRISTATE_CELLS:
- sky130_fd_sc_hd__ebuf*
VDD_NETS: null
VDD_PIN: VPWR
VDD_PIN_VOLTAGE: 1.8
WELLTAP_CELL: sky130_fd_sc_hd__tapvpwrvgnd_1
WIRE_LENGTH_THRESHOLD: null
meta:
  flow: null
  openlane_version: 2.3.10
  step: null
  substituting_steps: null
  version: 1

```


Next, we will synthesise the circuit.

In [9]:
Synthesis = Step.factory.get("Yosys.Synthesis")
synthesis = Synthesis(
    VERILOG_FILES=["/content/inverter32.v"],
    state_in=State(),
)
synthesis.start()

──────────────────────────────────────────────────── Synthesis ────────────────────────────────────────────────────

[16:10:58] VERBOSE  Running 'Yosys.Synthesis' at 'openlane_run/1-yosys-synthesis'…                     ]8;id=648118;file:///content/openlane_ipynb/openlane/steps/step.py\step.py]8;;\:]8;id=77779;file:///content/openlane_ipynb/openlane/steps/step.py#1122\1122]8;;\

[16:10:58] VERBOSE  Logging subprocess to 'openlane_run/1-yosys-synthesis/yosys-synthesis.log'…        ]8;id=64576;file:///content/openlane_ipynb/openlane/steps/step.py\step.py]8;;\:]8;id=439982;file:///content/openlane_ipynb/openlane/steps/step.py#1318\1318]8;;\

/----------------------------------------------------------------------------\

|  yosys -- Yosys Open SYnthesis Suite                                       |

|  Copyright (C) 2012 - 2024  Claire Xenia Wolf <claire@yosyshq.com>         |

|  Distributed under an ISC-like license, type "license" to see terms        |

\----------------------------------------------------------------------------/

Yosys 0.46 (git sha1 e97731b9dda91fa5fa53ed87df7c34163ba59a41, clang++ 17.0.6 -fPIC -O3)

Loaded SDC plugin

1. Executing Liberty frontend: /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lib/sky130_fd_sc_hd__tt_025C_1v80.lib

Imported 428 cell types from liberty file.

[INFO] Using SDC file '/content/openlane_run/1-yosys-synthesis/synthesis.abc.sdc' for ABC…

2. Executing Verilog-2005 frontend: /content/inverter32.v

Parsing SystemVerilog input from `/content/inverter32.v' to AST representation.

Storing AST representation for module `$abstract\inverter32'.

Successfully finished Verilog frontend.

3. Executing HIERARCHY pass (managing design hierarchy).

4. Executing AST frontend in derive mode using pre-parsed AST for module `\inverter32'.

Generating RTLIL representation for module `\inverter32'.

4.1. Analyzing design hierarchy..

Top module:  \inverter32

4.2. Analyzing design hierarchy..

Top module:  \inverter32

Removing unused module `$abstract\inverter32'.

Removed 1 unused modules.

Renaming module inverter32 to inverter32.

5. Generating Graphviz representation of design.

Writing dot description to `/content/openlane_run/1-yosys-synthesis/hierarchy.dot'.

Dumping module inverter32 to page 1.

6. Executing TRIBUF pass.

7. Executing HIERARCHY pass (managing design hierarchy).

7.1. Analyzing design hierarchy..

Top module:  \inverter32

7.2. Analyzing design hierarchy..

Top module:  \inverter32

Removed 0 unused modules.

8. Executing PROC_CLEAN pass (remove empty switches from decision trees).

Cleaned up 0 empty switches.

9. Executing PROC_RMDEAD pass (remove dead branches from decision trees).

Removed a total of 0 dead cases.

10. Executing PROC_PRUNE pass (remove redundant assignments in processes).

Removed 0 redundant assignments.

Promoted 0 assignments to connections.

11. Executing PROC_INIT pass (extract init attributes).

12. Executing PROC_ARST pass (detect async resets in processes).

13. Executing PROC_ROM pass (convert switches to ROMs).

Converted 0 switches.

14. Executing PROC_MUX pass (convert decision trees to multiplexers).

15. Executing PROC_DLATCH pass (convert process syncs to latches).

16. Executing PROC_DFF pass (convert process syncs to FFs).

17. Executing PROC_MEMWR pass (convert process memory writes to cells).

18. Executing PROC_CLEAN pass (remove empty switches from decision trees).

Cleaned up 0 empty switches.

19. Executing CHECK pass (checking for obvious problems).

Checking module inverter32...

Found and reported 0 problems.

20. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

21. Executing FLATTEN pass (flatten design).

22. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

23. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

Removed 0 unused cells and 1 unused wires.

<suppressed ~1 debug messages>

24. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

25. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

26. Executing OPT_MUXTREE pass (detect dead branches in mux trees).

Running muxtree optimizer on module \inverter32..

Creating internal representation of mux trees.

No muxes found in this module.

Removed 0 multiplexer ports.

27. Executing OPT_REDUCE pass (consolidate $*mux and $reduce_* inputs).

Optimizing cells in module \inverter32.

Performed a total of 0 changes.

28. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

29. Executing OPT_DFF pass (perform DFF optimizations).

30. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

31. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

32. Executing FSM pass (extract and optimize FSM).

32.1. Executing FSM_DETECT pass (finding FSMs in design).

32.2. Executing FSM_EXTRACT pass (extracting FSM from design).

32.3. Executing FSM_OPT pass (simple optimizations of FSMs).

32.4. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

32.5. Executing FSM_OPT pass (simple optimizations of FSMs).

32.6. Executing FSM_RECODE pass (re-assigning FSM state encoding).

32.7. Executing FSM_INFO pass (dumping all available information on FSM cells).

32.8. Executing FSM_MAP pass (mapping FSMs to basic logic).

33. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

34. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

35. Executing OPT_MUXTREE pass (detect dead branches in mux trees).

Running muxtree optimizer on module \inverter32..

Creating internal representation of mux trees.

No muxes found in this module.

Removed 0 multiplexer ports.

36. Executing OPT_REDUCE pass (consolidate $*mux and $reduce_* inputs).

Optimizing cells in module \inverter32.

Performed a total of 0 changes.

37. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

38. Executing OPT_DFF pass (perform DFF optimizations).

39. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

40. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

41. Executing WREDUCE pass (reducing word size of cells).

42. Executing PEEPOPT pass (run peephole optimizers).

43. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

44. Executing ALUMACC pass (create $alu and $macc cells).

Extracting $alu and $macc cells in module inverter32:

created 0 $alu and 0 $macc cells.

45. Executing SHARE pass (SAT-based resource sharing).

46. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

47. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

48. Executing OPT_MUXTREE pass (detect dead branches in mux trees).

Running muxtree optimizer on module \inverter32..

Creating internal representation of mux trees.

No muxes found in this module.

Removed 0 multiplexer ports.

49. Executing OPT_REDUCE pass (consolidate $*mux and $reduce_* inputs).

Optimizing cells in module \inverter32.

Performed a total of 0 changes.

50. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

51. Executing OPT_DFF pass (perform DFF optimizations).

52. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

53. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

54. Executing MEMORY pass.

54.1. Executing OPT_MEM pass (optimize memories).

Performed a total of 0 transformations.

54.2. Executing OPT_MEM_PRIORITY pass (removing unnecessary memory write priority relations).

Performed a total of 0 transformations.

54.3. Executing OPT_MEM_FEEDBACK pass (finding memory read-to-write feedback paths).

54.4. Executing MEMORY_BMUX2ROM pass (converting muxes to ROMs).

54.5. Executing MEMORY_DFF pass (merging $dff cells to $memrd).

54.6. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

54.7. Executing MEMORY_SHARE pass (consolidating $memrd/$memwr cells).

54.8. Executing OPT_MEM_WIDEN pass (optimize memories where all ports are wide).

Performed a total of 0 transformations.

54.9. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

54.10. Executing MEMORY_COLLECT pass (generating $mem cells).

55. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

56. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

57. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

58. Executing OPT_DFF pass (perform DFF optimizations).

59. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

60. Executing MEMORY_MAP pass (converting memories to logic and flip-flops).

61. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

62. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

63. Executing OPT_MUXTREE pass (detect dead branches in mux trees).

Running muxtree optimizer on module \inverter32..

Creating internal representation of mux trees.

No muxes found in this module.

Removed 0 multiplexer ports.

64. Executing OPT_REDUCE pass (consolidate $*mux and $reduce_* inputs).

Optimizing cells in module \inverter32.

Performed a total of 0 changes.

65. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

66. Executing OPT_SHARE pass.

67. Executing OPT_DFF pass (perform DFF optimizations).

68. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

69. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

70. Executing TECHMAP pass (map to technology primitives).

70.1. Executing Verilog-2005 frontend:                                                                             
/nix/store/9r0bh7sp051dpm8km8bqlb028anpd3v3-yosys/bin/../share/yosys/techmap.v

Parsing Verilog input from `/nix/store/9r0bh7sp051dpm8km8bqlb028anpd3v3-yosys/bin/../share/yosys/techmap.v' to AST 
representation.

Generating RTLIL representation for module `\_90_simplemap_bool_ops'.

Generating RTLIL representation for module `\_90_simplemap_reduce_ops'.

Generating RTLIL representation for module `\_90_simplemap_logic_ops'.

Generating RTLIL representation for module `\_90_simplemap_compare_ops'.

Generating RTLIL representation for module `\_90_simplemap_various'.

Generating RTLIL representation for module `\_90_simplemap_registers'.

Generating RTLIL representation for module `\_90_shift_ops_shr_shl_sshl_sshr'.

Generating RTLIL representation for module `\_90_shift_shiftx'.

Generating RTLIL representation for module `\_90_fa'.

Generating RTLIL representation for module `\_90_lcu_brent_kung'.

Generating RTLIL representation for module `\_90_alu'.

Generating RTLIL representation for module `\_90_macc'.

Generating RTLIL representation for module `\_90_alumacc'.

Generating RTLIL representation for module `\$__div_mod_u'.

Generating RTLIL representation for module `\$__div_mod_trunc'.

Generating RTLIL representation for module `\_90_div'.

Generating RTLIL representation for module `\_90_mod'.

Generating RTLIL representation for module `\$__div_mod_floor'.

Generating RTLIL representation for module `\_90_divfloor'.

Generating RTLIL representation for module `\_90_modfloor'.

Generating RTLIL representation for module `\_90_pow'.

Generating RTLIL representation for module `\_90_pmux'.

Generating RTLIL representation for module `\_90_demux'.

Generating RTLIL representation for module `\_90_lut'.

Successfully finished Verilog frontend.

70.2. Continuing TECHMAP pass.

Using extmapper simplemap for cells of type $not.

No more expansions possible.

<suppressed ~74 debug messages>

71. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

72. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

73. Executing OPT_DFF pass (perform DFF optimizations).

74. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

75. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

76. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

77. Executing OPT_DFF pass (perform DFF optimizations).

78. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

79. Executing ABC pass (technology mapping using ABC).

79.1. Extracting gate netlist of module `\inverter32' to `<abc-temp-dir>/input.blif'..

Extracted 32 gates and 64 wires to a netlist network with 32 inputs and 32 outputs.

79.1.1. Executing ABC.

Running ABC command: "/nix/store/6l3aj1gi4lja8z5s4lan2k2cfz9mqkrz-yosys-abc/bin/abc" -s -f                         
<abc-temp-dir>/abc.script 2>&1

ABC: ABC command line: "source <abc-temp-dir>/abc.script".

ABC:

ABC: + read_blif <abc-temp-dir>/input.blif

ABC: + read_library <abc-temp-dir>/stdcells.genlib

ABC: Entered genlib library with 13 gates from file "<abc-temp-dir>/stdcells.genlib".

ABC: + strash

ABC: + dretime

ABC: + map

ABC: + write_blif <abc-temp-dir>/output.blif

79.1.2. Re-integrating ABC results.

ABC RESULTS:               NOT cells:       32

ABC RESULTS:        internal signals:        0

ABC RESULTS:           input signals:       32

ABC RESULTS:          output signals:       32

Removing temp directory.

80. Executing OPT pass (performing simple optimizations).

80.1. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

80.2. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

80.3. Executing OPT_DFF pass (perform DFF optimizations).

80.4. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

Removed 0 unused cells and 64 unused wires.

<suppressed ~1 debug messages>

80.5. Finished fast OPT passes.

81. Executing HIERARCHY pass (managing design hierarchy).

81.1. Analyzing design hierarchy..

Top module:  \inverter32

81.2. Analyzing design hierarchy..

Top module:  \inverter32

Removed 0 unused modules.

82. Executing CHECK pass (checking for obvious problems).

Checking module inverter32...

Found and reported 0 problems.

83. Printing statistics.

=== inverter32 ===

Number of wires:                  2

Number of wire bits:             64

Number of public wires:           2

Number of public wire bits:      64

Number of ports:                  2

Number of port bits:             64

Number of memories:               0

Number of memory bits:            0

Number of processes:              0

Number of cells:                 32

$_NOT_                         32

84. Generating Graphviz representation of design.

Writing dot description to `/content/openlane_run/1-yosys-synthesis/primitive_techmap.dot'.

Dumping module inverter32 to page 1.

85. Executing OPT pass (performing simple optimizations).

85.1. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

85.2. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

85.3. Executing OPT_MUXTREE pass (detect dead branches in mux trees).

Running muxtree optimizer on module \inverter32..

Creating internal representation of mux trees.

No muxes found in this module.

Removed 0 multiplexer ports.

85.4. Executing OPT_REDUCE pass (consolidate $*mux and $reduce_* inputs).

Optimizing cells in module \inverter32.

Performed a total of 0 changes.

85.5. Executing OPT_MERGE pass (detect identical cells).

Finding identical cells in module `\inverter32'.

Removed a total of 0 cells.

85.6. Executing OPT_DFF pass (perform DFF optimizations).

85.7. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

85.8. Executing OPT_EXPR pass (perform const folding).

Optimizing module inverter32.

85.9. Finished OPT passes. (There is nothing left to do.)

86. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

{

"creator": "Yosys 0.46 (git sha1 e97731b9dda91fa5fa53ed87df7c34163ba59a41, clang++ 17.0.6 -fPIC -O3)",

"invocation": "stat -json -liberty /content/openlane_run/tmp/f87d52e94e2e4dc98f8beccf2759038c.lib ",

"modules": {

"\\inverter32": {

"num_wires":         2,

"num_wire_bits":     64,

"num_pub_wires":     2,

"num_pub_wire_bits": 64,

"num_ports":         2,

"num_port_bits":     64,

"num_memories":      0,

"num_memory_bits":   0,

"num_processes":     0,

"num_cells":         32,

"num_cells_by_type": {

"$_NOT_": 32

}

}

},

"design": {

"num_wires":         2,

"num_wire_bits":     64,

"num_pub_wires":     2,

"num_pub_wire_bits": 64,

"num_ports":         2,

"num_port_bits":     64,

"num_memories":      0,

"num_memory_bits":   0,

"num_processes":     0,

"num_cells":         32,

"num_cells_by_type": {

"$_NOT_": 32

}

}

}

87. Printing statistics.

=== inverter32 ===

Number of wires:                  2

Number of wire bits:             64

Number of public wires:           2

Number of public wire bits:      64

Number of ports:                  2

Number of port bits:             64

Number of memories:               0

Number of memory bits:            0

Number of processes:              0

Number of cells:                 32

$_NOT_                         32

Area for cell type $_NOT_ is unknown!

[INFO] Applying tri-state buffer mapping from                                                                      
'/root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/tribuff_map.v'…

88. Executing TECHMAP pass (map to technology primitives).

88.1. Executing Verilog-2005 frontend: /root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/tribuff_map.v

Parsing Verilog input from `/root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/tribuff_map.v' to AST         
representation.

Generating RTLIL representation for module `\$_TBUF_'.

Successfully finished Verilog frontend.

88.2. Continuing TECHMAP pass.

No more expansions possible.

<suppressed ~3 debug messages>

89. Executing SIMPLEMAP pass (map simple cells to gate primitives).

[INFO] Applying latch mapping from '/root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/latch_map.v'…

90. Executing TECHMAP pass (map to technology primitives).

90.1. Executing Verilog-2005 frontend: /root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/latch_map.v

Parsing Verilog input from `/root/.volare/sky130A/libs.tech/openlane/sky130_fd_sc_hd/latch_map.v' to AST           
representation.

Generating RTLIL representation for module `\$_DLATCH_P_'.

Generating RTLIL representation for module `\$_DLATCH_N_'.

Successfully finished Verilog frontend.

90.2. Continuing TECHMAP pass.

No more expansions possible.

<suppressed ~4 debug messages>

91. Executing SIMPLEMAP pass (map simple cells to gate primitives).

92. Executing DFFLIBMAP pass (mapping DFF cells to sequential cells from liberty file).

cell sky130_fd_sc_hd__dfxtp_2 (noninv, pins=3, area=21.27) is a direct match for cell type $_DFF_P_.

cell sky130_fd_sc_hd__dfrtp_2 (noninv, pins=4, area=26.28) is a direct match for cell type $_DFF_PN0_.

cell sky130_fd_sc_hd__dfstp_2 (noninv, pins=4, area=26.28) is a direct match for cell type $_DFF_PN1_.

cell sky130_fd_sc_hd__dfbbn_2 (noninv, pins=6, area=35.03) is a direct match for cell type $_DFFSR_NNN_.

final dff cell mappings:

unmapped dff cell: $_DFF_N_

\sky130_fd_sc_hd__dfxtp_2 _DFF_P_ (.CLK( C), .D( D), .Q( Q));

unmapped dff cell: $_DFF_NN0_

unmapped dff cell: $_DFF_NN1_

unmapped dff cell: $_DFF_NP0_

unmapped dff cell: $_DFF_NP1_

\sky130_fd_sc_hd__dfrtp_2 _DFF_PN0_ (.CLK( C), .D( D), .Q( Q), .RESET_B( R));

\sky130_fd_sc_hd__dfstp_2 _DFF_PN1_ (.CLK( C), .D( D), .Q( Q), .SET_B( R));

unmapped dff cell: $_DFF_PP0_

unmapped dff cell: $_DFF_PP1_

\sky130_fd_sc_hd__dfbbn_2 _DFFSR_NNN_ (.CLK_N( C), .D( D), .Q( Q), .Q_N(~Q), .RESET_B( R), .SET_B( S));

unmapped dff cell: $_DFFSR_NNP_

unmapped dff cell: $_DFFSR_NPN_

unmapped dff cell: $_DFFSR_NPP_

unmapped dff cell: $_DFFSR_PNN_

unmapped dff cell: $_DFFSR_PNP_

unmapped dff cell: $_DFFSR_PPN_

unmapped dff cell: $_DFFSR_PPP_

92.1. Executing DFFLEGALIZE pass (convert FFs to types supported by the target).

Mapping DFF cells in module `\inverter32':

{

"creator": "Yosys 0.46 (git sha1 e97731b9dda91fa5fa53ed87df7c34163ba59a41, clang++ 17.0.6 -fPIC -O3)",

"invocation": "stat -json -liberty /content/openlane_run/tmp/f87d52e94e2e4dc98f8beccf2759038c.lib ",

"modules": {

"\\inverter32": {

"num_wires":         2,

"num_wire_bits":     64,

"num_pub_wires":     2,

"num_pub_wire_bits": 64,

"num_ports":         2,

"num_port_bits":     64,

"num_memories":      0,

"num_memory_bits":   0,

"num_processes":     0,

"num_cells":         32,

"num_cells_by_type": {

"$_NOT_": 32

}

}

},

"design": {

"num_wires":         2,

"num_wire_bits":     64,

"num_pub_wires":     2,

"num_pub_wire_bits": 64,

"num_ports":         2,

"num_port_bits":     64,

"num_memories":      0,

"num_memory_bits":   0,

"num_processes":     0,

"num_cells":         32,

"num_cells_by_type": {

"$_NOT_": 32

}

}

}

93. Printing statistics.

=== inverter32 ===

Number of wires:                  2

Number of wire bits:             64

Number of public wires:           2

Number of public wire bits:      64

Number of ports:                  2

Number of port bits:             64

Number of memories:               0

Number of memory bits:            0

Number of processes:              0

Number of cells:                 32

$_NOT_                         32

Area for cell type $_NOT_ is unknown!

[INFO] Using generated ABC script '/content/openlane_run/1-yosys-synthesis/AREA_0.abc'…

94. Executing ABC pass (technology mapping using ABC).

94.1. Extracting gate netlist of module `\inverter32' to `/tmp/yosys-abc-sWEtVF/input.blif'..

Extracted 32 gates and 64 wires to a netlist network with 32 inputs and 32 outputs.

94.1.1. Executing ABC.

Running ABC command: "/nix/store/6l3aj1gi4lja8z5s4lan2k2cfz9mqkrz-yosys-abc/bin/abc" -s -f                         
/tmp/yosys-abc-sWEtVF/abc.script 2>&1

ABC: ABC command line: "source /tmp/yosys-abc-sWEtVF/abc.script".

ABC:

ABC: + read_blif /tmp/yosys-abc-sWEtVF/input.blif

ABC: + read_lib -w /content/openlane_run/tmp/f87d52e94e2e4dc98f8beccf2759038c.lib

ABC: Parsing finished successfully.  Parsing time =     0.12 sec

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfbbn_2".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfrbp_2".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfrtp_2".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfrtp_4".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfsbp_2".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfstp_2".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfstp_4".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfxbp_2".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfxtp_2".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dfxtp_4".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dlxtn_1".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dlxtn_2".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dlxtn_4".

ABC: Scl_LibertyReadGenlib() skipped sequential cell "sky130_fd_sc_hd__dlxtp_1".

ABC: Scl_LibertyReadGenlib() skipped three-state cell "sky130_fd_sc_hd__ebufn_2".

ABC: Scl_LibertyReadGenlib() skipped three-state cell "sky130_fd_sc_hd__ebufn_4".

ABC: Scl_LibertyReadGenlib() skipped three-state cell "sky130_fd_sc_hd__ebufn_8".

ABC: Library "sky130_fd_sc_hd__tt_025C_1v80" from "/content/openlane_run/tmp/f87d52e94e2e4dc98f8beccf2759038c.lib" 
has 175 cells (17 skipped: 14 seq; 3 tri-state; 0 no func; 0 dont_use).  Time =     0.17 sec

ABC: Memory =    9.54 MB. Time =     0.17 sec

ABC: Warning: Detected 2 multi-output gates (for example, "sky130_fd_sc_hd__fa_1").

ABC: + read_constr -v /content/openlane_run/1-yosys-synthesis/synthesis.abc.sdc

ABC: Setting driving cell to be "sky130_fd_sc_hd__inv_2/Y".

ABC: Setting output load to be 33.442001.

ABC: + source /content/openlane_run/1-yosys-synthesis/AREA_0.abc

ABC: Error: The network is combinational.

ABC: Cannot find the default PI driving cell (sky130_fd_sc_hd__inv_2/Y) in the library.

ABC: WireLoad = "none"  Gates =     32 (100.0 %)   Cap = 19.0 ff ( 12.5 %)   Area =      120.12 (  0.0 %)   Delay =
118.84 ps  (100.0 %)

ABC: Path  0 --       1 : 0    1 pi                     A =   0.00  Df =   0.0   -0.0 ps  S =   0.0 ps  Cin =  0.0 
ff  Cout =   4.6 ff  Cmax =   0.0 ff  G =    0

ABC: Path  1 --      65 : 1    1 sky130_fd_sc_hd__inv_2 A =   3.75  Df = 118.8  -57.1 ps  S = 157.7 ps  Cin =  4.5 
ff  Cout =  33.4 ff  Cmax = 331.4 ff  G =  749

ABC: Start-point = pi0 (\a_in [21]).  End-point = po0 (\y_out [21]).

ABC: netlist                       : i/o =   32/   32  lat =    0  nd =    32  edge =     32  area =120.00  delay =
1.00  lev = 1

ABC: + write_blif /tmp/yosys-abc-sWEtVF/output.blif

94.1.2. Re-integrating ABC results.

ABC RESULTS:   sky130_fd_sc_hd__inv_2 cells:       32

ABC RESULTS:        internal signals:        0

ABC RESULTS:           input signals:       32

ABC RESULTS:          output signals:       32

Removing temp directory.

95. Executing SETUNDEF pass (replace undef values with defined constants).

96. Executing HILOMAP pass (mapping to constant drivers).

97. Executing SPLITNETS pass (splitting up multi-bit signals).

98. Executing OPT_CLEAN pass (remove unused cells and wires).

Finding unused cells or wires in module \inverter32..

Removed 0 unused cells and 64 unused wires.

<suppressed ~1 debug messages>

99. Executing INSBUF pass (insert buffer cells for connected wires).

100. Executing CHECK pass (checking for obvious problems).

Checking module inverter32...

Found and reported 0 problems.

{

"creator": "Yosys 0.46 (git sha1 e97731b9dda91fa5fa53ed87df7c34163ba59a41, clang++ 17.0.6 -fPIC -O3)",

"invocation": "stat -json -liberty /content/openlane_run/tmp/f87d52e94e2e4dc98f8beccf2759038c.lib ",

"modules": {

"\\inverter32": {

"num_wires":         2,

"num_wire_bits":     64,

"num_pub_wires":     2,

"num_pub_wire_bits": 64,

"num_ports":         2,

"num_port_bits":     64,

"num_memories":      0,

"num_memory_bits":   0,

"num_processes":     0,

"num_cells":         32,

"area":              120.115200,

"num_cells_by_type": {

"sky130_fd_sc_hd__inv_2": 32

}

}

},

"design": {

"num_wires":         2,

"num_wire_bits":     64,

"num_pub_wires":     2,

"num_pub_wire_bits": 64,

"num_ports":         2,

"num_port_bits":     64,

"num_memories":      0,

"num_memory_bits":   0,

"num_processes":     0,

"num_cells":         32,

"area":              120.115200,

"num_cells_by_type": {

"sky130_fd_sc_hd__inv_2": 32

}

}

}

101. Printing statistics.

=== inverter32 ===

Number of wires:                  2

Number of wire bits:             64

Number of public wires:           2

Number of public wire bits:      64

Number of ports:                  2

Number of port bits:             64

Number of memories:               0

Number of memory bits:            0

Number of processes:              0

Number of cells:                 32

sky130_fd_sc_hd__inv_2         32

Chip area for module '\inverter32': 120.115200

of which used for sequential elements: 0.000000 (0.00%)

102. Executing Verilog backend.

Dumping module `\inverter32'.

103. Executing JSON backend.

[16:11:02] VERBOSE  Parsing synthesis checks…                                                          ]8;id=26771;file:///content/openlane_ipynb/openlane/steps/pyosys.py\pyosys.py]8;;\:]8;id=118290;file:///content/openlane_ipynb/openlane/steps/pyosys.py#54\54]8;;\

Format,Path
nl,openlane_run/1-yosys-synthesis/inverter32.nl.v


As we expected, there are 32 inverter standard cells. How many public wires and public wire bits are there? Can you guess their meaning?

* Number of public wires: 2
* Number of wire bits: 64

Public wires(2) relates to the input and output cable for each cell, with wire bits relating to the total count of individual 1-bit boolean signal lines (wires).

Next, we will proceed with floorplanning. Remember that the tool is set to find the smallest possible size of the die.

In [10]:
Floorplan = Step.factory.get("OpenROAD.Floorplan")
floorplan = Floorplan(state_in=synthesis.state_out)
floorplan.start()

──────────────────────────────────────────── Floorplan Initialization ─────────────────────────────────────────────

[16:11:54] VERBOSE  Running 'OpenROAD.Floorplan' at 'openlane_run/2-openroad-floorplan'…               ]8;id=50024;file:///content/openlane_ipynb/openlane/steps/step.py\step.py]8;;\:]8;id=640753;file:///content/openlane_ipynb/openlane/steps/step.py#1122\1122]8;;\

[16:11:54] VERBOSE  Logging subprocess to 'openlane_run/2-openroad-floorplan/openroad-floorplan.log'…  ]8;id=94867;file:///content/openlane_ipynb/openlane/steps/step.py\step.py]8;;\:]8;id=941486;file:///content/openlane_ipynb/openlane/steps/step.py#1318\1318]8;;\

Reading library file at '/root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lib/sky130_fd_sc_hd__tt_025C_1v80.lib'…

Reading technology LEF file at '/root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/techlef/sky130_fd_sc_hd__nom.tlef'…

[INFO ODB-0227] LEF file: /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/techlef/sky130_fd_sc_hd__nom.tlef, created
14 layers, 25 vias

Reading cell LEF file at '/root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_fd_sc_hd.lef'…

The NOWIREEXTENSIONATPIN statement will be ignored. See file                                                       
/root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_fd_sc_hd.lef at line 2.

[INFO ODB-0227] LEF file: /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_fd_sc_hd.lef, created 437      
library cells

Reading cell LEF file at '/root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_ef_sc_hd.lef'…

The NOWIREEXTENSIONATPIN statement will be ignored. See file                                                       
/root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_ef_sc_hd.lef at line 2.

[INFO ODB-0227] LEF file: /root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_ef_sc_hd.lef, created 4 library
cells

Reading top-level netlist at '/content/openlane_run/1-yosys-synthesis/inverter32.nl.v'…

Linking design 'inverter32' from netlist…

Reading design constraints file at '/content/openlane_ipynb/openlane/scripts/base.sdc'…

[16:11:56] WARNING  [STA-0366] port '__VIRTUAL_CLK__' not found.                                    ]8;id=339604;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=988221;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Using clock __VIRTUAL_CLK__…

[INFO] Setting output delay to: 2

[INFO] Setting input delay to: 2

[16:11:56] WARNING  [STA-0366] port '__VIRTUAL_CLK__' not found.                                    ]8;id=672312;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=974555;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Setting load to: 0.033442

[INFO] Setting clock uncertainty to: 0.25

[INFO] Setting clock transition to: 0.1499999999999999944488848768742172978818416595458984375

[16:11:56] WARNING  [STA-0419] transition time can not be specified for virtual clocks.             ]8;id=332812;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=651797;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Setting timing derate to: 5%

[16:11:56] WARNING  [STA-0450] virtual clock __VIRTUAL_CLK__ can not be propagated.                 ]8;id=780063;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=286152;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

Using site height: 2.72 and site width: 0.46…

[INFO] Using relative sizing for the floorplan.

[INFO IFP-0001] Added 5 rows of 33 site unithd.

[INFO IFP-0030] Inserted 0 tiecells using sky130_fd_sc_hd__conb_1/LO.

[INFO IFP-0030] Inserted 0 tiecells using sky130_fd_sc_hd__conb_1/HI.

[INFO] Extracting DIE_AREA and CORE_AREA from the floorplan

[INFO] Floorplanned on a die area of 0.0 0.0 26.54 37.26 (µm).

[INFO] Floorplanned on a core area of 5.52 10.88 20.7 24.48 (µm).

Writing metric design__die__bbox: 0.0 0.0 26.54 37.26

Writing metric design__core__bbox: 5.52 10.88 20.7 24.48

Setting global connections for newly added cells…

[INFO] Setting global connections...

Updating metrics…

Cell type report:                       Count       Area

Inverter                                 32     120.12

Total                                    32     120.12

Writing OpenROAD database to '/content/openlane_run/2-openroad-floorplan/inverter32.odb'…

Writing netlist to '/content/openlane_run/2-openroad-floorplan/inverter32.nl.v'…

Writing powered netlist to '/content/openlane_run/2-openroad-floorplan/inverter32.pnl.v'…

Writing layout to '/content/openlane_run/2-openroad-floorplan/inverter32.def'…

Writing timing constraints to '/content/openlane_run/2-openroad-floorplan/inverter32.sdc'…

Format,Path
nl,openlane_run/2-openroad-floorplan/inverter32.nl.v
pnl,openlane_run/2-openroad-floorplan/inverter32.pnl.v
def,openlane_run/2-openroad-floorplan/inverter32.def
odb,openlane_run/2-openroad-floorplan/inverter32.odb
sdc,openlane_run/2-openroad-floorplan/inverter32.sdc


Look for die area in the log. What is the dimension of the die?

The die dimensions are: 26.54 µm (width) x 37.26 µm (height)

Next, we will insert tap cells and end cap cells. Then randomly assign I/O pins.

In [11]:
TapEndcapInsertion = Step.factory.get("OpenROAD.TapEndcapInsertion")
tdi = TapEndcapInsertion(state_in=floorplan.state_out)
tdi.start()

IOPlacement = Step.factory.get("OpenROAD.IOPlacement")
ioplace = IOPlacement(state_in=tdi.state_out)
ioplace.start()
display(ioplace)

─────────────────────────────────────────────── Tap/Decap Insertion ───────────────────────────────────────────────

[16:15:26] VERBOSE  Running 'OpenROAD.TapEndcapInsertion' at                                           ]8;id=534027;file:///content/openlane_ipynb/openlane/steps/step.py\step.py]8;;\:]8;id=982466;file:///content/openlane_ipynb/openlane/steps/step.py#1122\1122]8;;\
                    'openlane_run/3-openroad-tapendcapinsertion'…                                                  

[16:15:26] VERBOSE  Logging subprocess to                                                              ]8;id=591617;file:///content/openlane_ipynb/openlane/steps/step.py\step.py]8;;\:]8;id=326465;file:///content/openlane_ipynb/openlane/steps/step.py#1318\1318]8;;\
                    'openlane_run/3-openroad-tapendcapinsertion/openroad-tapendcapinsertion.log'…                  

Reading OpenROAD database at '/content/openlane_run/2-openroad-floorplan/inverter32.odb'…

Reading library file at '/root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lib/sky130_fd_sc_hd__tt_025C_1v80.lib'…

Reading design constraints file at '/content/openlane_ipynb/openlane/scripts/base.sdc'…

[16:15:26] WARNING  [STA-0366] port '__VIRTUAL_CLK__' not found.                                    ]8;id=53609;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=412659;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Using clock __VIRTUAL_CLK__…

[INFO] Setting output delay to: 2

[INFO] Setting input delay to: 2

[16:15:26] WARNING  [STA-0366] port '__VIRTUAL_CLK__' not found.                                    ]8;id=848743;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=166575;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Setting load to: 0.033442

[INFO] Setting clock uncertainty to: 0.25

[INFO] Setting clock transition to: 0.1499999999999999944488848768742172978818416595458984375

[16:15:26] WARNING  [STA-0419] transition time can not be specified for virtual clocks.             ]8;id=597817;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=183123;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Setting timing derate to: 5%

[16:15:26] WARNING  [STA-0450] virtual clock __VIRTUAL_CLK__ can not be propagated.                 ]8;id=969165;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=73624;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO TAP-0004] Inserted 10 endcaps.

[INFO TAP-0005] Inserted 3 tapcells.

Setting global connections for newly added cells…

[INFO] Setting global connections...

Updating metrics…

Cell type report:                       Count       Area

Fill cell                                10      37.54

Tap cell                                  3       3.75

Inverter                                 32     120.12

Total                                    45     161.40

Writing OpenROAD database to '/content/openlane_run/3-openroad-tapendcapinsertion/inverter32.odb'…

Writing netlist to '/content/openlane_run/3-openroad-tapendcapinsertion/inverter32.nl.v'…

Writing powered netlist to '/content/openlane_run/3-openroad-tapendcapinsertion/inverter32.pnl.v'…

Writing layout to '/content/openlane_run/3-openroad-tapendcapinsertion/inverter32.def'…

Writing timing constraints to '/content/openlane_run/3-openroad-tapendcapinsertion/inverter32.sdc'…

────────────────────────────────────────────────── I/O Placement ──────────────────────────────────────────────────

[16:15:27] VERBOSE  Running 'OpenROAD.IOPlacement' at 'openlane_run/4-openroad-ioplacement'…           ]8;id=502036;file:///content/openlane_ipynb/openlane/steps/step.py\step.py]8;;\:]8;id=429670;file:///content/openlane_ipynb/openlane/steps/step.py#1122\1122]8;;\

[16:15:27] VERBOSE  Logging subprocess to                                                              ]8;id=92490;file:///content/openlane_ipynb/openlane/steps/step.py\step.py]8;;\:]8;id=10124;file:///content/openlane_ipynb/openlane/steps/step.py#1318\1318]8;;\
                    'openlane_run/4-openroad-ioplacement/openroad-ioplacement.log'…                                

Reading OpenROAD database at '/content/openlane_run/3-openroad-tapendcapinsertion/inverter32.odb'…

Reading library file at '/root/.volare/sky130A/libs.ref/sky130_fd_sc_hd/lib/sky130_fd_sc_hd__tt_025C_1v80.lib'…

Reading design constraints file at '/content/openlane_ipynb/openlane/scripts/base.sdc'…

[16:15:27] WARNING  [STA-0366] port '__VIRTUAL_CLK__' not found.                                    ]8;id=325433;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=545122;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Using clock __VIRTUAL_CLK__…

[INFO] Setting output delay to: 2

[INFO] Setting input delay to: 2

[16:15:27] WARNING  [STA-0366] port '__VIRTUAL_CLK__' not found.                                    ]8;id=388898;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=217801;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Setting load to: 0.033442

[INFO] Setting clock uncertainty to: 0.25

[INFO] Setting clock transition to: 0.1499999999999999944488848768742172978818416595458984375

[16:15:27] WARNING  [STA-0419] transition time can not be specified for virtual clocks.             ]8;id=732233;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=227572;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] Setting timing derate to: 5%

[16:15:27] WARNING  [STA-0450] virtual clock __VIRTUAL_CLK__ can not be propagated.                 ]8;id=954234;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=2317;file:///content/openlane_ipynb/openlane/steps/openroad.py#235\235]8;;\

[INFO] place_pins args: -min_distance 3

Found 0 macro blocks.

[16:15:27] ERROR    [PPL-0024] Number of IO pins (64) exceeds maximum number of available positions ]8;id=136299;file:///content/openlane_ipynb/openlane/steps/openroad.py\openroad.py]8;;\:]8;id=147067;file:///content/openlane_ipynb/openlane/steps/openroad.py#233\233]8;;\
                    (40). Increase the die perimeter from 127.60um to 217.60um.                                    

Error: ioplacer.tcl, 62 PPL-0024

StepError: OpenROAD.IOPlacement failed with the following errors:
[PPL-0024] Number of IO pins (64) exceeds maximum number of available positions (40). Increase the die perimeter from 127.60um to 217.60um.

You should receive an error at this box. What is the explanation of the **ERROR** from the floorplan?

This error occurs during the pin placement stage (PPL-0024) in OpenLane because the design requires more input/output (I/O) pins than can physically fit along the current perimeter of the floorplanned die, therefore the die perimeter is required to increase from 127.60 µm to 217.60 µm.

## Step 3: Second trial – floorplan with fixed die size

Based on the suggestion from the tools, provide a new dimension for the die to fit all I/O pins. Enter your answers (new width, new height) in the form below.



In [ ]:
from openlane.steps import Step
from openlane.state import State
from openlane.config import Config

new_die_width = 10 # @param {key:"New die width", type:"number"}
new_die_height = 10 # @param {key:"New die height", type:"number"}
Config.interactive(
    "inverter32",
    PDK="sky130A",
    DIE_AREA=[0, 0, new_die_width, new_die_height],
    PRIMARY_GDSII_STREAMOUT_TOOL="klayout",
)

Let's synthesise again.

In [ ]:
Synthesis = Step.factory.get("Yosys.Synthesis")
synthesis = Synthesis(
    VERILOG_FILES=["/content/inverter32.v"],
    state_in=State(),
)
synthesis.start()

This time, we ask the floorplanner to respect the die size supplied with the option ``FP_SIZING`` as **absolute**. Run the floorplanning again.

In [ ]:
Floorplan = Step.factory.get("OpenROAD.Floorplan")
floorplan = Floorplan(FP_SIZING="absolute", state_in=synthesis.state_out)
floorplan.start()

Inspect the output on die area and confirm that the new dimension is in place.

What is the core area in μm${}^2$?

*Enter your answer here*

Next, we will try again with the I/O placement.

In [ ]:
TapEndcapInsertion = Step.factory.get("OpenROAD.TapEndcapInsertion")
tdi = TapEndcapInsertion(state_in=floorplan.state_out)
tdi.start()

IOPlacement = Step.factory.get("OpenROAD.IOPlacement")
ioplace = IOPlacement(state_in=tdi.state_out)
ioplace.start()
display(ioplace)

If it is **not successful**, return to the first box in **Step 3** and enter a bigger width and height.

Which metal layer(s) is/are used for the I/O pins?

*Enter your answer here*

Estimate the pitch of the I/O pins. The pitch is the distance between two pins.

*Enter your answer here*

We will proceed with the remaining steps of the physical implementation.

In [ ]:
# generate power distribution network (PDN)
GeneratePDN = Step.factory.get("OpenROAD.GeneratePDN")
pdn = GeneratePDN(
    state_in=ioplace.state_out,
    FP_PDN_AUTO_ADJUST=True,
    FP_PDN_VPITCH=25,
    FP_PDN_HPITCH=25,
    FP_PDN_VOFFSET=5,
    FP_PDN_HOFFSET=5,
)
pdn.start()

# Global placement
GlobalPlacement = Step.factory.get("OpenROAD.GlobalPlacement")
gpl = GlobalPlacement(state_in=pdn.state_out)
gpl.start()

# run detailed placement
DetailedPlacement = Step.factory.get("OpenROAD.DetailedPlacement")
dpl = DetailedPlacement(state_in=gpl.state_out)
dpl.start()

display(dpl)

Next, we will route the design.

In [ ]:
# Global routing
GlobalRouting = Step.factory.get("OpenROAD.GlobalRouting")
grt = GlobalRouting(state_in=dpl.state_out)
grt.start()

DetailedRouting = Step.factory.get("OpenROAD.DetailedRouting")
drt = DetailedRouting(state_in=grt.state_out)
drt.start()
display(drt)

Inspect the log and the routed design. What is the final total wirelength?

*Enter your answer here*

You may consider this as the length required to connect the logic circuit with the I/O pins.

Please save a copy of the picture of this final layout for your final assignment.

## Step 4: Challenges of inserting I/O pads

From the task above, we have successfully assigned I/O pins around the die, but without any I/O pads. I/O pads are necessary to provide electrical protection and an area for wire bonding.

A typical in-line pad pitch required for 0.18 μm technology is [70 μm](https://www.onsemi.com/PowerSolutions/content.do?id=16679). What would be the dimension of the die if in-line pads are added?

*Enter your answer here*

How many percentage of the area is used as the core in this case?

*Enter your answer here*


---

**What did I learn?**

When you have completed the manual floorplanning to fit in I/O pins and pads, add your thoughts in response to the following question. You will want to keep these, as you will find them useful when working on later modules and your final assignment. To get started, double click this cell and edit the text.

**How does assembly and packaging affect chip design? What are the necessary steps during design to prepare the chip for final packaging?**

*Enter your answer here.*


---
You have now reached the end of the notebook for this week. Please return to Canvas to complete the rest of the work in module 5.
